src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs

```diff
@@ -388,83 +367,94 @@ public int Compare(string? string1, int offset1, string? string2, int offset2, C
+            return Compare(string1, offset1, string2, offset2, CompareOptions.None);
         }
 
         public int Compare(string? string1, int offset1, int length1, string? string2, int offset2, int length2, CompareOptions options)
         {
-            if (options == CompareOptions.OrdinalIgnoreCase)
+            ReadOnlySpan<char> span1 = default;
+            ReadOnlySpan<char> span2 = default;
+
+            if (string1 == null)
             {
-                int result = string.Compare(string1, offset1, string2, offset2, length1 < length2 ? length1 : length2, StringComparison.OrdinalIgnoreCase);
-                if ((length1 != length2) && result == 0)
+                if (offset1 != 0 || length1 != 0)
                 {
-                    return length1 > length2 ? 1 : -1;
+                    goto BoundsCheckError;
                 }
-
-                return result;
-            }
-
-            if (length1 < 0 || length2 < 0)
-            {
-                throw new ArgumentOutOfRangeException((length1 < 0) ? nameof(length1) : nameof(length2), SR.ArgumentOutOfRange_NeedPosNum);
             }
-            if (offset1 < 0 || offset2 < 0)
-            {
-                throw new ArgumentOutOfRangeException((offset1 < 0) ? nameof(offset1) : nameof(offset2), SR.ArgumentOutOfRange_NeedPosNum);
-            }
-            if (offset1 > (string1 == null ? 0 : string1.Length) - length1)
-            {
-                throw new ArgumentOutOfRangeException(nameof(string1), SR.ArgumentOutOfRange_OffsetLength);
-            }
-            if (offset2 > (string2 == null ? 0 : string2.Length) - length2)
+            else if (!string1.TryGetSpan(offset1, length1, out span1))
             {
-                throw new ArgumentOutOfRangeException(nameof(string2), SR.ArgumentOutOfRange_OffsetLength);
+                goto BoundsCheckError;
             }
-            if ((options & CompareOptions.Ordinal) != 0)
+
+            if (string2 == null)
             {
-                if (options != CompareOptions.Ordinal)
+                if (offset2 != 0 || length2 != 0)
                 {
-                    throw new ArgumentException(SR.Argument_CompareOptionOrdinal,
-                                                nameof(options));
+                    goto BoundsCheckError;
                 }
             }
-            else if ((options & ValidCompareMaskOffFlags) != 0)
+            else if (!string2.TryGetSpan(offset2, length2, out span2))
             {
-                throw new ArgumentException(SR.Argument_InvalidFlag, nameof(options));
+                goto BoundsCheckError;
             }
 
+            // At this point both string1 and string2 have been bounds-checked.
+
+            int retVal;
+
+            // Our paradigm is that null sorts less than any other string and
+            // that two nulls sort as equal.
+
             if (string1 == null)
             {
```

Comment:
didn't we already checked string == 1 before?

Reply:
It would've ended up duplicating the _string2_ bounds checking within the `if (string1 == null) { /* ... */ }` branch. Seemed cleanest to perform all of the parameter checking first, then short-circuit nulls once the bounds checks finished.